File: packages/tutoriales/src/motor.ts
import { Injectable, computed, inject, signal } from '@angular/core'
import { Router } from '@angular/router'
import { BITACORA_DE_TUTORIALES } from './bitacora'
import { cerrado, enPaso, progresoInicial } from './avance'
import { EncendidoDeTutorial } from './encendido'
import { LocalizadorDeObjetivo } from './objetivo'
import { ProgresoDeTutoriales } from './progreso'
import { RegistroDeTutoriales } from './registro'
import type { PasoDeTutorial, ProblemaEnEjecucion, ProgresoDeTutorial, TutorialDefinicion } from './tipos'

/**
 * **El motor: dónde estoy, cómo sigo y qué hago si el paso no se puede dar.**
 *
 * Es una máquina de estados con señales; no dibuja nada. El overlay, el globo y el
 * teclado viven en `layout/tutoriales/`, y el catálogo en `rutas/ayuda/catalogo/`.
 * Entre los tres no hay ningún `if` con el id de un tutorial concreto.
 *
 * Tres reglas que no se negocian:
 * 1. **No se traba.** Un objetivo que no aparece es un problema declarado con reintento
 *    y salida, nunca una pantalla congelada.
 * 2. **No opera.** Lo único que ejecuta solo es desplazar y enfocar. Ni un clic.
 * 3. **No inventa permisos.** Los pasos ya vienen recortados por `RegistroDeTutoriales`.
 */
@Injectable({ providedIn: 'root' })
export class MotorDeTutoriales {
  private readonly registro = inject(RegistroDeTutoriales)
  private readonly localizador = inject(LocalizadorDeObjetivo)
  private readonly progresos = inject(ProgresoDeTutoriales)
  private readonly encendido = inject(EncendidoDeTutorial)
  private readonly bitacora = inject(BITACORA_DE_TUTORIALES)
  private readonly router = inject(Router)

  private readonly tutorialActivo = signal<TutorialDefinicion | null>(null)
  private readonly indiceActual = signal(0)
  private readonly progresoActual = signal<ProgresoDeTutorial | null>(null)

  readonly tutorial = this.tutorialActivo.asReadonly()
  readonly indice = this.indiceActual.asReadonly()
  readonly activo = computed(() => this.tutorialActivo() !== null)
  readonly paso = computed<PasoDeTutorial | null>(() => this.tutorialActivo()?.pasos[this.indiceActual()] ?? null)
  readonly total = computed(() => this.tutorialActivo()?.pasos.length ?? 0)
  readonly esUltimo = computed(() => this.indiceActual() >= this.total() - 1)

  /** El elemento que hay que resaltar. `null` es «todavía no» o «no está». */
  readonly elemento = signal<HTMLElement | null>(null)
  /** Lo que impide seguir, en palabras. Se muestra en el globo, con reintento. */
  readonly problema = signal<ProblemaEnEjecucion | null>(null)
  /** La acción que el paso pedía ya está hecha. */
  readonly accionCumplida = signal(false)
  /** Está preguntando si de verdad quiere abandonar a la mitad. */
  readonly confirmandoSalida = signal(false)
  /** No pudimos guardar el avance (cuota, modo privado). Se avisa, no se esconde. */
  readonly sinGuardar = signal(false)

  readonly puedeAvanzar = computed(() => this.paso()?.accion === undefined || this.paso()?.accion?.tipo === 'ninguna' || this.accionCumplida())

  /** Empieza de cero. Si ya se había hecho, suma repetición al completarlo. */
  async iniciar(tutorialId: string, repeticiones = 0): Promise<void> {
    const tutorial = this.registro.buscar(tutorialId)
    if (tutorial === undefined || tutorial.pasos.length === 0) return
    this.encender(tutorial)
    this.progresoActual.set(progresoInicial(tutorial, ahora(), repeticiones))
    this.bitacora.anotar({ tipo: 'inicio', tutorialId })
    await this.ir(0)
  }

  /** Retoma donde quedó. Un índice fuera de rango vuelve al principio. */
  async continuar(tutorialId: string, progreso: ProgresoDeTutorial): Promise<void> {
    const tutorial = this.registro.buscar(tutorialId)
    if (tutorial === undefined || tutorial.pasos.length === 0) return
    this.encender(tutorial)
    this.progresoActual.set(progreso)
    this.bitacora.anotar({ tipo: 'inicio', tutorialId })
    await this.ir(Math.min(Math.max(progreso.indice, 0), tutorial.pasos.length - 1))
  }

  async avanzar(): Promise<void> {
    if (!this.puedeAvanzar()) {
      const paso = this.paso()
      if (paso !== null) this.anotarProblema('accion-pendiente', paso, paso.ayudaSiFalla ?? 'Hacé lo que pide el paso para seguir.')
      return
    }
    if (this.esUltimo()) return this.terminar()
    await this.ir(this.indiceActual() + 1)
  }

  async retroceder(): Promise<void> {
    if (this.indiceActual() === 0) return
    await this.ir(this.indiceActual() - 1)
  }

  /** Reintenta el paso actual: vuelve a buscar el objetivo desde cero. */
  async reintentar(): Promise<void> {
    await this.ir(this.indiceActual())
  }

  /** Pide confirmación si va por la mitad; si recién empezó, cierra sin preguntar. */
  pedirSalida(): void {
    if (this.indiceActual() === 0 || this.esUltimo()) return void this.omitir()
    this.confirmandoSalida.set(true)
  }

  seguirEnElTutorial(): void {
    this.confirmandoSalida.set(false)
  }

  /** Abandona: queda como omitido, con el paso donde se fue. Se puede repetir después. */
  omitir(): void {
    const tutorial = this.tutorialActivo()
    const progreso = this.progresoActual()
    if (tutorial !== null && progreso !== null) {
      this.bitacora.anotar({ tipo: 'omitido', tutorialId: tutorial.id, pasoId: this.paso()?.id ?? null })
      this.guardar(cerrado(enPaso(progreso, this.indiceActual(), this.paso()?.id ?? null, ahora()), 'omitido', ahora()))
    }
    this.apagar()
  }

  private async terminar(): Promise<void> {
    const tutorial = this.tutorialActivo()
    const progreso = this.progresoActual()
    if (tutorial !== null && progreso !== null) {
      this.bitacora.anotar({ tipo: 'fin', tutorialId: tutorial.id })
      this.guardar(cerrado({ ...progreso, indice: tutorial.pasos.length }, 'completado', ahora()))
    }
    this.apagar()
  }

  /**
   * El corazón: navegar si el paso vive en otra pantalla, esperar al elemento y dejar
   * todo listo. Cada regreso temprano deja un problema declarado, nunca silencio.
   */
  private async ir(indice: number): Promise<void> {
    const tutorial = this.tutorialActivo()
    const paso = tutorial?.pasos[indice]
    if (tutorial === undefined || tutorial === null || paso === undefined) return
    this.indiceActual.set(indice)
    this.accionCumplida.set(false)
    this.problema.set(null)
    this.elemento.set(null)
    this.bitacora.anotar({ tipo: 'paso', tutorialId: tutorial.id, pasoId: paso.id, indice })
    const progreso = this.progresoActual()
    if (progreso !== null) this.guardar(enPaso(progreso, indice, paso.id, ahora()))

    // El primer paso hereda la ruta del tutorial: `ruta` de la definición es «dónde
    // empieza». Sin esto, abrir desde el centro de ayuda un tutorial de otra pantalla
    // dejaba a la persona en el centro, con el primer objetivo imposible de encontrar.
    if (!(await this.llegarA(paso.ruta ?? (indice === 0 ? tutorial.ruta : undefined)))) return
    if (paso.objetivo === undefined) return
    const elemento = await this.localizador.esperar(paso.objetivo, paso.esperaMs)
    if (elemento === null) {
      this.anotarProblema('sin-objetivo', paso, `No encontramos «${paso.objetivo}» en esta pantalla.`)
      return
    }
    this.elemento.set(elemento)
    if (paso.automatica !== undefined) this.localizador.acercar(elemento, paso.automatica === 'enfocar')
  }

  /** Lleva a esa ruta si hace falta. `false` si no se pudo. */
  private async llegarA(ruta: string | undefined): Promise<boolean> {
    if (ruta === undefined) return true
    const actual = this.router.url.split('?')[0]
    if (actual === ruta || actual?.startsWith(`${ruta}/`) === true) return true
    const llegó = await this.router.navigateByUrl(ruta).catch(() => false)
    const paso = this.paso()
    if (llegó === false && paso !== null) {
      this.anotarProblema('sin-ruta', paso, `No pudimos abrir «${ruta}». Puede que tu rol no tenga esa sección.`)
      return false
    }
    return llegó !== false
  }

  private anotarProblema(motivo: ProblemaEnEjecucion['motivo'], paso: PasoDeTutorial, detalle: string): void {
    this.problema.set({ motivo, pasoId: paso.id, detalle })
    const tutorialId = this.tutorialActivo()?.id
    if (tutorialId !== undefined) this.bitacora.anotar({ tipo: 'problema', tutorialId, pasoId: paso.id, detalle })
  }

  private guardar(progreso: ProgresoDeTutorial): void {
    this.progresoActual.set(progreso)
    this.progresos.guardar(progreso).subscribe({ next: () => this.sinGuardar.set(false), error: () => this.sinGuardar.set(true) })
  }

  private encender(tutorial: TutorialDefinicion): void {
    this.tutorialActivo.set(tutorial)
    this.encendido.activo.set(true)
  }

  private apagar(): void {
    this.tutorialActivo.set(null)
    this.encendido.activo.set(false)
    this.elemento.set(null)
    this.problema.set(null)
    this.confirmandoSalida.set(false)
    this.indiceActual.set(0)
    this.progresoActual.set(null)
  }
}

const ahora = (): string => new Date().toISOString()

File: apps/movil/lib/dominio/tutoriales/motor.dart
import 'package:aportaya_diseno/moviles/anclas_de_tutorial.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../proveedores/tutoriales.dart';
import 'avance.dart';
import 'escribano.dart';
import 'estado_del_motor.dart';
import 'modelo.dart';

/// **El motor: dónde estoy, cómo sigo y qué hago si el paso no se puede dar.**
///
/// Es una máquina de estados; no dibuja nada. El velo, el globo y los gestos viven en
/// `pantallas/soporte/capa_de_tutorial.dart`, y el catálogo en `pantallas/soporte/
/// catalogo/`. Entre los tres no hay un solo `if` con el id de un tutorial concreto.
///
/// Tres reglas que no se negocian:
/// 1. **No se traba.** Un objetivo que no aparece es un problema declarado con
///    reintento y salida, nunca una pantalla congelada.
/// 2. **No opera.** No toca botones, no envía formularios, no confirma nada.
/// 3. **No inventa permisos.** Los tutoriales ya vienen filtrados por capacidad.
class MotorDeTutoriales extends Notifier<EstadoDelMotor> {
  late final _escribano = EscribanoDeProgreso(ref);
  AnclasDeTutorial? _anclas;
  void Function(String ruta)? _navegar;
  String Function()? _ubicacion;

  @override
  EstadoDelMotor build() => EstadoDelMotor.apagado;

  /// La capa visual le presta al motor lo que solo existe con un `BuildContext`: el
  /// registro de anclas y la navegación. Se llama una vez, al montarse la capa.
  void conectar({
    required AnclasDeTutorial anclas,
    required void Function(String ruta) navegar,
    required String Function() ubicacion,
  }) {
    _anclas = anclas;
    _navegar = navegar;
    _ubicacion = ubicacion;
  }

  /// Empieza de cero. Si ya se había hecho, las repeticiones viajan.
  Future<void> iniciar(String tutorialId, {int repeticiones = 0}) async {
    final tutorial = ref.read(registroDeTutorialesProvider).buscar(tutorialId);
    if (tutorial == null || tutorial.pasos.isEmpty) return;
    state = EstadoDelMotor(tutorial: tutorial);
    await _escribano.guardar(
      progresoInicial(tutorial, DateTime.now(), repeticiones: repeticiones),
    );
    await _ir(0);
  }

  /// Retoma donde quedó. Un índice fuera de rango vuelve al último paso posible.
  Future<void> continuar(String tutorialId, ProgresoDeTutorial progreso) async {
    final tutorial = ref.read(registroDeTutorialesProvider).buscar(tutorialId);
    if (tutorial == null || tutorial.pasos.isEmpty) return;
    state = EstadoDelMotor(tutorial: tutorial);
    await _ir(progreso.indice.clamp(0, tutorial.pasos.length - 1));
  }

  Future<void> avanzar() async {
    if (!state.puedeAvanzar) {
      state = state.copiaCon(
        problema:
            state.paso?.ayudaSiFalla ?? 'Hacé lo que pide el paso para seguir.',
      );
      return;
    }
    if (state.esUltimo) return _terminar();
    await _ir(state.indice + 1);
  }

  Future<void> retroceder() async {
    if (state.indice == 0) return;
    await _ir(state.indice - 1);
  }

  /// Reintenta el paso actual: vuelve a buscar el objetivo desde cero.
  Future<void> reintentar() => _ir(state.indice);

  /// La acción que el paso pedía ya está hecha.
  void marcarAccionCumplida() {
    if (state.activo) state = state.copiaCon(accionCumplida: true);
  }

  /// Pide confirmación si va por la mitad; si recién empezó, cierra sin preguntar.
  void pedirSalida() {
    if (state.indice == 0 || state.esUltimo) {
      omitir();
      return;
    }
    state = state.copiaCon(confirmandoSalida: true);
  }

  void seguirEnElTutorial() {
    state = state.copiaCon(confirmandoSalida: false);
  }

  /// Abandona: queda omitido, con el paso donde se fue. Se puede repetir después.
  void omitir() {
    final tutorial = state.tutorial;
    if (tutorial != null) {
      final ahora = DateTime.now();
      _escribano.sinEsperar(
        cerrado(
          enPaso(
            _escribano.de(tutorial, ahora),
            state.indice,
            state.paso?.id,
            ahora,
          ),
          EstadoDeProgreso.omitido,
          ahora,
        ),
      );
    }
    state = EstadoDelMotor.apagado;
  }

  Future<void> _terminar() async {
    final tutorial = state.tutorial;
    state = EstadoDelMotor.apagado;
    if (tutorial == null) return;
    final ahora = DateTime.now();
    final hecho = _escribano
        .de(tutorial, ahora)
        .copiaCon(indice: tutorial.pasos.length);
    await _escribano.guardar(
      cerrado(hecho, EstadoDeProgreso.completado, ahora),
    );
  }

  /// El corazón: navegar si el paso vive en otra pantalla, esperar al elemento y dejar
  /// todo listo. Cada regreso temprano deja un problema declarado, nunca silencio.
  Future<void> _ir(int indice) async {
    final tutorial = state.tutorial;
    if (tutorial == null || indice >= tutorial.pasos.length) return;
    final paso = tutorial.pasos[indice];
    state = state.copiaCon(
      indice: indice,
      accionCumplida: false,
      limpiarProblema: true,
      limpiarRecuadro: true,
    );

    _escribano.sinEsperar(
      enPaso(
        _escribano.de(tutorial, DateTime.now()),
        indice,
        paso.id,
        DateTime.now(),
      ),
    );

    // El primer paso hereda la ruta del tutorial: `ruta` de la definición es «dónde
    // empieza». Sin esto, abrir desde Ayuda un tutorial de otra pantalla dejaba a la
    // persona en Ayuda, con el primer ancla imposible de encontrar.
    if (!_llegarA(paso.ruta ?? (indice == 0 ? tutorial.ruta : null))) return;
    if (paso.ancla == null) return;
    final recuadro = await _anclas?.esperar(paso.ancla!, paso.espera);
    if (!state.activo || state.indice != indice) return;
    if (recuadro == null) {
      state = state.copiaCon(
        problema: 'No encontramos «${paso.ancla}» en esta pantalla.',
      );
      return;
    }
    state = state.copiaCon(recuadro: recuadro);
  }

  /// Lleva a esa ruta si hace falta. `false` si no se pudo.
  bool _llegarA(String? ruta) {
    if (ruta == null) return true;
    final actual = _ubicacion?.call() ?? '';
    if (actual == ruta || actual.startsWith('$ruta/')) return true;
    if (_navegar == null) {
      state = state.copiaCon(problema: 'No pudimos abrir «$ruta» desde acá.');
      return false;
    }
    _navegar!(ruta);
    return true;
  }

  /// Vuelve a medir el elemento del paso actual: la pantalla se desplazó o cambió de
  /// tamaño y el agujero tiene que seguirla.
  void medirDeNuevo() {
    final ancla = state.paso?.ancla;
    if (!state.activo || ancla == null) return;
    final recuadro = _anclas?.recuadroDe(ancla);
    if (recuadro != null && recuadro != state.recuadro) {
      state = state.copiaCon(recuadro: recuadro);
    }
  }
}
